src/libraries/System.Private.CoreLib/src/System/String.Comparison.cs

```diff
@@ -834,43 +836,108 @@ internal unsafe int GetNonRandomizedHashCode()
             }
         }
 
-        // Use this if and only if 'Denial of Service' attacks are not a concern (i.e. never used for free-form user input),
-        // or are otherwise mitigated
-        internal unsafe int GetNonRandomizedHashCodeOrdinalIgnoreCase()
+        internal int GetNonRandomizedHashCodeOrdinalIgnoreCase()
         {
-            fixed (char* src = &_firstChar)
-            {
-                Debug.Assert(src[this.Length] == '\0', "src[this.Length] == '\\0'");
-                Debug.Assert(((int)src) % 4 == 0, "Managed string should start at 4 bytes boundary");
+            return GetNonRandomizedHashCodeOrdinalIgnoreCaseStatic(ref _firstChar, Length, true);
+        }
 
-                uint hash1 = (5381 << 16) + 5381;
-                uint hash2 = hash1;
+        private static int GetNonRandomizedHashCodeOrdinalIgnoreCaseStatic(ref char firstChar, int length, bool normalizeNonAscii)
+        {
+            Debug.Assert(Unsafe.Add(ref firstChar, length) == '\0', "*(&firstChar + length) == '\\0'");
```

Comment:
I am not fan of byref pointer arithmetic that is non-measurable improvement.

Reply:
neither am I, but I thought you guys prefer byref to avoid pinning everywhere. will rewrite.